candid_rust/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+    } else {
+        idl_type_usage_to_rust(
+            context,
+            &func_t.rets.first().unwrap(),
+            &Span::call_site(),
+            false,
+        )?
+    };
+
+    let return_type = if func_t.modes.contains(&FuncMode::Query) {
+        return_type
+    } else {
+        quote! { impl std::future::Future<Output = #return_type> }
+    };
+
+    Ok(quote! {
+        fn #id(#(#arguments_def),*) -> #return_type
+    })
+}
+
+fn idl_type_declaration_to_rust(
+    context: &Context,
+    ty: &Dec,
+    span: &dyn Spanned,
+) -> Result<TokenStream> {
+    let (id_str, typ) = match ty {
+        Dec::ImportD(_) => Err(errors::Error::Custom("Imports not supported.".to_string())),
+        Dec::TypD(binding) => Ok((&binding.id, &binding.typ)),
+    }?;
+
+    let id = syn::Ident::new(id_str, span.span());
+
+    // We separate variants, structs and enums from the rest, as the rest can be used
+    // as `type X = #idl_type_usage_to_rust()`
+    // but idl_type_usage_to_rust() errors on some types because they're invalid in a
+    // type usage in rust.
+    return Ok(match typ {
+        IDLType::RecordT(fields) => {
+            let mut all_fields = Vec::new();
+            for f in fields {
+                let label_str = match &f.label {
+                    Label::Id(i) => format!("id_{}", i),
+                    Label::Named(n) => n.clone(),
+                    Label::Unnamed(i) => format!("unnamed_{}", i),
+                };
+                let label = syn::Ident::new(&label_str, span.span());
+                let ty = idl_type_usage_to_rust(context, &f.typ, &span.span(), true)?;
+
+                all_fields.push(quote! {
+                    pub #label : #ty
+                });
+            }
+
+            quote! {
+                struct #id {
+                    #(#all_fields),*
+                }
+            }
+        }
+        ty => {
+            let usage = idl_type_usage_to_rust(context, ty, span, false)?;
+            quote! { type #id = #usage ; }
+        }
+    });
+}
+
+fn idl_type_usage_to_rust(
+    context: &Context,
+    ty: &IDLType,
+    span: &dyn Spanned,
+    should_box: bool,
+) -> Result<TokenStream> {
+    Ok(match ty {
+        IDLType::PrimT(prim_t) => match prim_t {
+            // TODO: support bigint.
+            PrimType::Nat => quote! { u128 },
+            PrimType::Nat8 => quote! { u8 },
+            PrimType::Nat16 => quote! { u16 },
+            PrimType::Nat32 => quote! { u32 },
+            PrimType::Nat64 => quote! { u64 },
+            PrimType::Int => quote! { i128 },
+            PrimType::Int8 => quote! { i8 },
+            PrimType::Int16 => quote! { i16 },
+            PrimType::Int32 => quote! { i32 },
+            PrimType::Int64 => quote! { i64 },
+            PrimType::Float32 => quote! { f32 },
+            PrimType::Float64 => quote! { f64 },
+            PrimType::Bool => quote! { bool },
+            PrimType::Text => quote! { String },
```

Comment:
Do we have a freedom to choose between `String` and `&str`? #justsaying

Reply:
Not if it's in a struct :/

Reply:
It's possible with a lifetime in a struct...

Reply:
Yes but that complicates the generation and I'd like to keep it simple.